File: vista_comic/vista_comic/Features/Study/Scheduler.swift
//
//  Scheduler.swift
//  vista_comic
//
//  What one answer does to a card — the same state machine the backend runs,
//  in Swift.
//
//  **A deliberate second implementation.** Everywhere else in this app the rule
//  is that the backend decides and the app is told, precisely so the two cannot
//  come to disagree. Offline practice cannot ask, so this exists, and the cost
//  is paid down two ways: the transition table below is the same table
//  `backend/app/scheduler.py` documents, and `SchedulerParityTests` asserts the
//  same cases against both wordings so a change to one that is not made to the
//  other fails a test rather than a reader's schedule.
//
//  When an answer does reach the server, the server's result **wins** — this is
//  what the app runs on until then, not a second opinion to reconcile.
//

import Foundation

/// Everything scheduling knows about one card, as a value.
struct CardScheduling: Hashable {
    var state: CardState
    var learningStep: Int?
    var stage: Int
    var previousStage: Int?
    var dueAt: Date
}

/// Where `current` goes after one answer.
///
/// | From | Correct | Wrong |
/// |---|---|---|
/// | `new` | learning, step 0 | learning, step 0 |
/// | learning step *i* | step *i+1* | step 0 |
/// | learning, last step | graduates | step 0 |
/// | `review` slot *n* | slot *n+1*, clamped | relearning, keeping *n−1* |
/// | `relearning` | as learning | step 0 |
///
/// A `review` card answered **before** its due date is returned unchanged in
/// either direction — see the comment on that branch.
///
/// A first answer that is wrong still only reaches the first step: there is
/// nothing below the bottom, and inventing a punishment there would charge the
/// reader for meeting a word.
///
/// **The learning steps land on a minute; the slots land on a day.** A step
/// means its minutes literally. A slot means a *day* rather than a block of
/// twenty-four hours, so those due dates are rounded to the start of a
/// scheduling day (`dueAfter`) — which is what stops a card finished at
/// 23:59 from coming back at 23:59 the following night.
///
/// `timeZone` decides whose day that is. It defaults to the phone's, which is
/// the right answer for every caller here: all three are "this reader, on this
/// device, now". The backend is configured with a fixed zone instead
/// (`config.get_scheduling_timezone`), and the two agree at home; where they
/// do not, the server's result wins on sync, as it does for everything else.
///
/// Graduating returns to `previousStage` when there is one, which is what makes
/// a lapse cost one slot instead of everything — and it is cleared once used,
/// so a second lapse from slot 5 returns to 4 rather than to whatever the first
/// remembered.
func nextSchedule(
    _ current: CardScheduling,
    correct: Bool,
    answeredAt: Date,
    learningSteps: [Int],
    timeZone: TimeZone = .current
) -> CardScheduling {
    let steps = learningSteps.filter { $0 > 0 }
    // The backend raises on an unusable list because a route can answer 422.
    // Here there is nobody to tell, and refusing to schedule would strand the
    // card — so the defaults stand in.
    let usable = steps.isEmpty ? StudySettings.fallback.learningSteps : steps

    func atStep(_ state: CardState, _ index: Int) -> CardScheduling {
        // Clamped, because the reader may have shortened the list under a card
        // already past the new end of it.
        let bounded = max(0, min(index, usable.count - 1))
        var moved = current
        moved.state = state
        moved.learningStep = bounded
        moved.dueAt = answeredAt.addingTimeInterval(TimeInterval(usable[bounded]) * 60)
        return moved
    }

    func graduated() -> CardScheduling {
        let slot = clampSlot(current.previousStage ?? 0)
        return CardScheduling(
            state: .review,
            learningStep: nil,
            stage: slot,
            previousStage: nil,
            dueAt: dueAfter(slot: slot, from: answeredAt, in: timeZone)
        )
    }

    switch current.state {
    case .new:
        return atStep(.learning, 0)

    case .learning, .relearning:
        if !correct { return atStep(current.state, 0) }
        let index = max(0, min(current.learningStep ?? 0, usable.count - 1))
        return index >= usable.count - 1 ? graduated() : atStep(current.state, index + 1)

    case .review:
        // **An answer that arrives before the card is due moves nothing** —
        // not up, and by the same sentence not down either. The interval is
        // the claim being tested, and a card that graduated onto one day an
        // hour ago has not survived a day, so nothing it can be answered with
        // now is a measurement of three.
        //
        // It is also what stops a card being asked twice from counting twice.
        // The queue is built from a deck this app holds, and a deck can be
        // stale; the card's own due date cannot.
        //
        // The learning steps are deliberately outside this: `learnAheadWindow`
        // offers a learning card up to twenty minutes early on purpose.
        if answeredAt < current.dueAt { return current }

        if correct {
            let slot = clampSlot(current.stage + 1)
            return CardScheduling(
                state: .review,
                learningStep: nil,
                stage: slot,
                previousStage: nil,
                dueAt: dueAfter(slot: slot, from: answeredAt, in: timeZone)
            )
        }
        var lapsed = atStep(.relearning, 0)
        lapsed.previousStage = clampSlot(current.stage - 1)
        return lapsed
    }
}

/// A slot brought inside the interval table.
func clampSlot(_ slot: Int) -> Int {
    max(0, min(slot, ladderTopRung))
}

/// The hour a scheduling day begins, in the reader's own timezone.
///
/// Four in the morning rather than midnight — Anki's default, for Anki's
/// reason. This reader reads comics at night, and under a midnight rollover a
/// card finished at 23:59 would come back **one minute later**, in the same
/// sitting. A day that ends when the reader goes to sleep is the day they
/// actually live.
///
/// Mirrored in `backend/app/ladder.py` as `DAY_ROLLOVER_HOUR`.
let ladderDayRolloverHour = 4

/// The start of the scheduling day `moment` falls in.
///
/// A scheduling day runs from `ladderDayRolloverHour` to the same hour the
/// next calendar day, so an answer given in the small hours belongs to the day
/// the reader still thinks of as in progress — 02:00 on Tuesday is Monday
/// night.
func schedulingDayStart(of moment: Date, in timeZone: TimeZone = .current) -> Date {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    guard let rollover = calendar.date(
        bySettingHour: ladderDayRolloverHour, minute: 0, second: 0, of: moment
    ) else {
        // Only reachable if the zone has no such hour on this date at all.
        // Falling back to the calendar day keeps a schedule rather than
        // refusing one, which is the same stance `nextSchedule` takes on an
        // unusable list of steps.
        return calendar.startOfDay(for: moment)
    }
    guard moment < rollover else { return rollover }
    return calendar.date(byAdding: .day, value: -1, to: rollover) ?? rollover
}

/// When a card on `slot` answered at `moment` comes back.
///
/// Counted from the answer rather than from the previous due date: a card
/// answered four days late is not owed those four days back.
///
/// **Measured in days, not in multiples of 24 hours.** The answer is rounded
/// down to the start of its scheduling day and the interval added to that, so
/// "one day" means "next day" whether the answer came at breakfast or at
/// 23:59. Added with `Calendar` rather than as seconds so a zone that observes
/// DST still lands on the rollover hour on the far side of a clock change.
func dueAfter(slot: Int, from moment: Date, in timeZone: TimeZone = .current) -> Date {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    let start = schedulingDayStart(of: moment, in: timeZone)
    let days = ladderIntervals[clampSlot(slot)]
    return calendar.date(byAdding: .day, value: days, to: start)
        ?? start.addingTimeInterval(TimeInterval(days) * 86_400)
}

extension LearningCard {
    /// This card's schedule, as a value the transition function can take.
    var scheduling: CardScheduling {
        CardScheduling(
            state: state,
            learningStep: learningStep,
            stage: ladderStage,
            previousStage: previousStage,
            dueAt: dueAt
        )
    }

    /// Takes a locally computed schedule, exactly as `apply(_:)` takes the
    /// server's. `introducedOn` is set here rather than sent, because the day a
    /// card was met is the reader's day and the app is the side that knows it.
    mutating func apply(_ scheduling: CardScheduling, introducedOn day: String) {
        if state == .new && scheduling.state != .new { introducedOn = day }
        state = scheduling.state
        learningStep = scheduling.learningStep
        ladderStage = scheduling.stage
        previousStage = scheduling.previousStage
        dueAt = scheduling.dueAt
    }
}

File: backend/app/scheduler.py
"""What one answer does to a card: the state machine, as a pure function.

The model is Anki's, which is SM-2's, with this project's binary grading. A card
walks a short list of **learning steps** measured in minutes, and on clearing
the last one it **graduates** onto ``ladder``'s interval table, measured in
days. A graduated card answered wrong walks the steps again and comes back one
slot lower than it was.

**It replaces the three-step day**, which was derived by replaying a day's
answers and reset at midnight -- so a card left half-learned when the reader put
the phone down was back at first sight the next morning, and their two answers
had bought nothing. That reset is the mechanism behind the first acceptance
pass's "I have practised this card several times and it still says New".

Nothing here reads a clock. ``answered_at`` is a parameter because an answer
given in airplane mode arrives hours after it happened, and scheduling it from
the moment it arrived would put a five-minute step in the wrong afternoon.
Settings are parameters for the same reason: the reader can change them, and a
function that reached for the current ones could not be tested against a
three-step list and a five-step list in the same run.
"""

from __future__ import annotations

from dataclasses import dataclass, replace
from datetime import datetime, timedelta
from enum import Enum
from typing import Optional, Sequence

from . import ladder

#: Minutes between learning steps, before a card graduates. Adjustable by the
#: reader (stage 6 ticket 06); this is the default, and Anki's own default of
#: 1m/10m was widened because a session here is one sitting rather than a queue
#: the reader grinds down.
DEFAULT_LEARNING_STEPS = (5, 7, 10)

#: How many cards may be met for the first time in one day, by default.
DEFAULT_NEW_CARDS_PER_DAY = 15


class CardState(str, Enum):
    """Where a card is in the model.

    Stored on the card rather than derived from its answers, which is a reversal
    of the three-step day's design and is forced by two things: a learning
    card's position is *which step, due at what minute*, a dimension the review
    log has never carried; and stage 6's migration resets every card while
    keeping every review row, so "has answers" stops meaning "has been met".
    """

    #: Never answered, or reset by a migration. Waits for the day's new-card
    #: quota rather than for a due date.
    NEW = "new"
    #: Walking the learning steps for the first time.
    LEARNING = "learning"
    #: Graduated, on the interval table.
    REVIEW = "review"
    #: Graduated once, then missed. Walking the same steps back to its slot.
    RELEARNING = "relearning"


@dataclass(frozen=True)
class CardSchedule:
    """Everything scheduling knows about one card.

    A value, not a row: the same shape comes out of the database, goes into
    ``next_schedule``, and comes back to be written. Keeping it separate from
    the ORM object is what lets the transitions be tested without a database.
    """

    state: CardState
    #: Index into the learning steps, or ``None`` outside the learning states.
    learning_step: Optional[int]
    #: Which slot of ``ladder.LADDER_INTERVALS``.
    stage: int
    #: The slot to return to on graduating from ``RELEARNING``. ``None`` when
    #: the card has never lapsed, or has already used it.
    previous_stage: Optional[int]
    due_at: datetime


def new_card(*, now: datetime) -> CardSchedule:
    """A freshly collected card: due immediately, waiting on the quota."""
    return CardSchedule(
        state=CardState.NEW,
        learning_step=None,
        stage=ladder.FIRST_SLOT,
        previous_stage=None,
        due_at=now,
    )


def next_schedule(
    current: CardSchedule,
    *,
    correct: bool,
    answered_at: datetime,
    learning_steps: Sequence[int] = DEFAULT_LEARNING_STEPS,
) -> CardSchedule:
    """Where ``current`` goes after one answer.

    ================  ===============================  ========================
    From              Correct                          Wrong
    ================  ===============================  ========================
    ``NEW``           learning, step 0                 learning, step 0
    learning step i   step i+1                         step 0
    learning, last    graduates                        step 0
    ``REVIEW`` n      slot n+1 (clamped at the top)    relearning, keeping n-1
    ``RELEARNING``    as learning                      step 0
    ================  ===============================  ========================

    **A ``REVIEW`` card answered before ``due_at`` is returned unchanged**, in
    either direction. See the comment on that branch: the interval is the claim
    being tested, and an answer given early has not tested it.

    **A first answer that is wrong still only reaches the first step.** There is
    nothing below the bottom to fall to, and inventing a punishment there would
    charge the reader for meeting a word.

    **Graduating goes to ``previous_stage`` when there is one**, which is what
    makes a lapse cost one slot instead of everything: 60 days -> missed -> the
    steps again -> 21 days, not 1. The old ladder fell all the way, which was
    defensible when a wrong answer was rare and is not now that judging is an
    exact match on production -- one word out of place in a Vietnamese sentence
    is wrong, and charging a year of progress for it teaches nothing.

    ``previous_stage`` is **cleared once used**, so a second lapse from slot 5
    returns to 4 rather than to whatever the first lapse remembered.
    """
    steps = tuple(learning_steps)
    if not steps or any(minutes <= 0 for minutes in steps):
        raise ValueError("learning steps must be a non-empty list of positive minutes")

    def at_step(state: CardState, index: int) -> CardSchedule:
        # Clamped, because the reader may have shortened the list under a card
        # already past the new end of it.
        bounded = max(0, min(index, len(steps) - 1))
        return replace(
            current,
            state=state,
            learning_step=bounded,
            due_at=answered_at + timedelta(minutes=steps[bounded]),
        )

    def graduated() -> CardSchedule:
        slot = ladder.clamp_slot(
            current.previous_stage
            if current.previous_stage is not None
            else ladder.FIRST_SLOT
        )
        return CardSchedule(
            state=CardState.REVIEW,
            learning_step=None,
            stage=slot,
            previous_stage=None,
            due_at=ladder.due_after(slot, answered_at),
        )

    if current.state is CardState.NEW:
        return at_step(CardState.LEARNING, 0)

    if current.state in (CardState.LEARNING, CardState.RELEARNING):
        if not correct:
            return at_step(current.state, 0)
        index = max(0, min(current.learning_step or 0, len(steps) - 1))
        if index >= len(steps) - 1:
            return graduated()
        return at_step(current.state, index + 1)

    # REVIEW.
    #
    # **An answer that arrives before the card is due moves nothing.** It is
    # recorded like any other -- what the reader did is a fact -- but it cannot
    # promote a slot, and by the same sentence it cannot cost one either: an
    # answer that is not allowed to earn anything must not be allowed to charge
    # anything.
    #
    # The interval is the claim being tested. A card that graduated onto "one
    # day" an hour ago has not survived a day, so answering it correctly now
    # says nothing about whether three days is the right next question, and
    # promoting it would teach the schedule something that did not happen.
    #
    # In practice this is also the guard against being asked twice. The queue
    # lives in the app and is built from a deck the app holds; a dropped
    # response leaves that deck believing a graduated card is still due, and it
    # will offer it again. The card's own due date is the only thing in the
    # system that can refuse.
    #
    # The learning steps are deliberately not covered: ``learnAheadWindow``
    # offers a learning card up to twenty minutes early **on purpose**, and a
    # duplicate there costs one step and heals itself.
    if answered_at < current.due_at:
        return current

    if correct:
        slot = ladder.clamp_slot(current.stage + 1)
        return CardSchedule(
            state=CardState.REVIEW,
            learning_step=None,
            stage=slot,
            previous_stage=None,
            due_at=ladder.due_after(slot, answered_at),
        )
    return replace(
        at_step(CardState.RELEARNING, 0),
        previous_stage=ladder.clamp_slot(current.stage - 1),
    )
